sailor/assetcentral/system.py

```diff
@@ -44,20 +44,85 @@ def get_property_mapping(cls):
             'template_id': ('templateID', None, None, None),
         }
 
+    @staticmethod
+    def _sort_components(comps):
+        """Sort components by order attribute.
+
+        Dictionary comps is not sorted by order number
+        Hence, we have to sort - we do this in a data frame
+        First, we sort by [model, order] to get order within model into column 'model_order'
+        Second, we sort by order to have the correct order in the component dictionary
+        """
+        name, order, id, model, row = [], [], [], [], []
+        r = 0
+        for c in comps['childNodes']:
+            name.append(c['name'])
+            order.append(c['order'])
+            id.append(c['id'])
+            model.append(c['model'])
+            row.append(r)
+            r += 1
+        compdf = pd.DataFrame(list(zip(name, id, order, model, row)), columns=['name', 'id', 'order', 'model', 'row'])
+        compdf['one'] = 1
+        compdf.sort_values(by=['model', 'order'], inplace=True)
+        compdf['model_order'] = compdf.groupby(['model']).agg(rank=('one', 'cumsum'))
+        compdf.sort_values(by='order', inplace=True)
+        return compdf
+
+    @staticmethod
+    def _traverse_components(comps, model_order, equipment_ids, system_ids):
+        """Traverse component structure recursively."""
+        compd = {}
+        compd['id'] = comps['id']
+        compd['name'] = comps['name']
+        if comps['order'] is not None:
+            if comps['objectType'] == 'EQU':
+                equipment_ids.append(compd['id'])
+            else:
+                system_ids.append(compd['id'])
+        compd['key'] = (comps['model'], model_order)
+        compd['object_type'] = comps['objectType']
+        compdf = System._sort_components(comps)
+        if len(compdf) > 0:
+            compd['child_nodes'] = []
+            for c in range(len(compdf)):
+                row = compdf.iloc[c]['row']
+                model_order = compdf.iloc[c]['model_order']
+                compd0, equipment_ids, system_ids = System._traverse_components(comps['childNodes'][row], model_order,
+                                                                                equipment_ids, system_ids)
+                compd['child_nodes'].append(compd0)
+        return compd, equipment_ids, system_ids
+
+    def _update_dictionary(self, comps):
+        if comps['object_type'] == 'EQU':
+            obj = self.equipments.filter(id=comps['id'])[0]
+            comps['indicators'] = self.indicators[comps['id']]
+        else:
+            if comps['id'] == self.id:
+                obj = self
+            else:
+                obj = self.systems.filter(id=comps['id'])[0]
+        comps['key'] = (obj.model_id, comps['key'][1])
+        if 'child_nodes' in comps.keys():
+            for c in comps['child_nodes']:
+                self._update_dictionary(c)
```

Comment:
Trying to understand the dictionary. Correct me if I'm wrong. And maybe some questions...

The top level system is the top level of the dict and it is identified by the `key` key.  The value of `child_nodes`  is a list of child nodes, where each child node is a dict of the same structure.

1. What is the second component of the key used for (the one you call `model_order`)?
2. How is the dictionary supposed to be used after all? Does it become clear if I read the hierarchy concept?